File: app/src/main/java/ladsoft/com/popularmoviesapp/core/mvp/presenter/MvpPresenter.java
package ladsoft.com.popularmoviesapp.core.mvp.presenter;

import java.lang.ref.WeakReference;

import ladsoft.com.popularmoviesapp.core.mvp.Mvp;

public abstract class MvpPresenter<V> implements Mvp.Presenter<V> {

    private WeakReference<V> view;

    @Override
    public void attachView(V view) {
        this.view = new WeakReference<>(view);
    }

    @Override
    public void detachView() {
        this.view = null;
    }

    @Override
    public boolean isViewAttached() {
        return this.view != null && this.view.get()!= null;
    }

    @Override
    public V getView() {
        if(!isViewAttached()) {
            throw new IllegalStateException("Presenter is in detached state");
        }

        return view.get();
    }

}

File: app/src/main/java/ladsoft/com/popularmoviesapp/core/mvp/presenter/MvpPresenter.kt
package ladsoft.com.popularmoviesapp.core.mvp.presenter

import java.lang.ref.WeakReference

import ladsoft.com.popularmoviesapp.core.mvp.Mvp

abstract class MvpPresenter<V> : Mvp.Presenter<V> {

    private var _view: WeakReference<V>? = null

    override fun attachView(view: V) {
        this._view = WeakReference(view)
    }

    override fun detachView() {
        this._view = null
    }

    override val isViewAttached: Boolean
        get() = this._view != null && this._view!!.get() != null


    override fun getView(): V? {
        if (!isViewAttached) {
            throw IllegalStateException("Presenter is in detached state")
        }

        return _view?.get()
    }

}
